### PR TITLE
Updating for C++17

### PR DESCRIPTION
**Pull request by sean-parent**
- Minimum requirement now C++14, supports C++17
- added `adobe::invoke_result_t` to polyfill C++17
- added `adobe::construct_at()` to polyfill C++20
- deprecated `adobe::construct()` -> use `adobe::construct_at()`
- Removed `std::binary_operation` and `std::unary_operation`
- Updated to boost 1.76.0
- Fixed section divider comments to all be 100 characters
- Reviewed and removed some #if 0 sections
- Misc minor improvements
- (re)ran clang-format
- Updated release version to 1.0.44

### CODE REVIEW

File: adobe/forest.hpp [hunk 1]
```diff
@@ -260,14 +253,16 @@ class reverse_fullorder_iterator
...
     typedef I iterator_type;
     typedef typename inherited_t::reference reference;
 
     reverse_fullorder_iterator() : edge_m(forest_trailing_edge) {}
-    explicit reverse_fullorder_iterator(I x) : base_m(--x), edge_m(forest_leading_edge - base_m.edge()) {}
+    explicit reverse_fullorder_iterator(I x)
+        : base_m(--x), edge_m(forest_leading_edge - base_m.edge()) {}
     template <typename U>
     reverse_fullorder_iterator(const reverse_fullorder_iterator<U>& x)
         : base_m(--x.base()), edge_m(forest_leading_edge - base_m.edge()) {}
```
Comment (fosterbrereton) [inline adobe/forest.hpp:268]: I think this is a bug; no need to decrement the iterator when it's already a `reverse_fullorder_iterator`
```suggestion
        : base_m(x.base()), edge_m(forest_leading_edge - base_m.edge()) {}
```

**Review by fosterbrereton — APPROVED** [review]
LGTM!

File: adobe/forest.hpp (hunk 1 above)
Comment (sean-parent) [reply adobe/forest.hpp:269]: Should we simply replace adobe::forest with:
```
namespace adobe {

using stlab::forest;
// and other associated types

}
```

That's how I handled the migration of `adobe::copy_on_write` -> `stlab::copy_on_write`

Comment (sean-parent) [reply adobe/forest.hpp:269]: Also - are you sure about the fix? `I` is a `fullorder_iterator`. I think you intended to change the next case.

Comment (fosterbrereton) [reply adobe/forest.hpp:269]: Yes, you're right. I intended to change the next case, not the one in the suggestion.

Comment (fosterbrereton) [reply adobe/forest.hpp:269]: I'm fine with ASL having a dependency on stlab; that seems like a pretty big change to me, though, but if we're already doing it for copy_on_write, then I think it should be the way to go in this case, too.

Comment (fosterbrereton) [reply adobe/forest.hpp:269]: As for the change, I think it's actually on the correct line, even though GitHub is highlighting another one. The suggestion shows line 269, which is the `reverse_fullorder_iterator` case.

**Comment by sean-parent** [comment]
I'm going to push this and open an issue to add the stlab dependency (already implicit for copy_on_write but not needed for building tests).

**Pull request merged**